### DISCUSSION TITLE
How to tighten the type of a schema

### DISCUSSION THREAD

**Discussion by xsjcTony** [colinhacks/zod#2781] [Q&A] [answered]
In `yup`, I can do something like this to tighten the type of a string schema simply:
```ts
type Foo = 'foo' | 'bar'

const a = yup.string<Foo>().required()
type A = yup.InferType<typeof a>
//   ^? Foo
```

But in `zod`, seems currently the only way I can achieve it is by `z.custom`? Is there any better alternative or more straight forward way to do so?
```ts
const b = z.custom<Foo>(str => typeof str === 'string' && ['foo', 'bar'].includes(str))
type B = z.infer<typeof b>
//   ^? Foo
```
And it's even worse if I want to make it optional... where I have to write more in the `check` function and invoke `optional()` just for the type, where it actually does nothing. Overall the whole thing seems weird.
```ts
const c = z.custom<Foo>(str => str === void 0 || typeof str === 'string' && ['foo', 'bar'].includes(str))
  .optional()
type C = z.infer<typeof c>
//   ^? Foo | undefined
```

There's actually another way, but way more chaos... (Look at that I need to write another `optional()` just for type inference which actually does nothing!)... This is a little bit disgusting to deal with. Compare to `yup`'s `yup.string<Foo>()`😥. But I believe there should be an easier way right?

```ts
const d = z.string().optional()
  .refine(val => val ? ['foo', 'bar'].includes(val) : true)
  // Here I have to write ".optional()" again otherwise it won't be in the final inferred type
  .pipe(z.custom<Foo>().optional());
type D = z.infer<typeof d>
//   ^? Foo | undefined
```

Also, the type of `check` function of `z.custom` is a little bit weird to me, like it's reasonable to make received `val` to be `unknown`, but why it's returning `any` instead of `boolean`? 

And also what's the usage of `.refinement()`? I can't find anything related in the documentation

---

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wJ4FcxwGZQQhwDk2YJAUKJLHAN5wBecAvvocSUxACZWVKMDGACmcAGIQIcALyk80knAA+pAEbIoAygGMIAO1TxkczDgB0xqMAMBzADxSIAPgAUASgtRRARyzAPryeQiLiAIJm5BYAkgZ4olAAKmEOwmIQeCgulAD0uXCFAHoA-IJ6hsZw6mZMFrpYxkRO0u7Wci5w6aKZcO2yA6TWtnbKAGRjcADaJIoQJAA0GlokALoWtroANli8oqhu1h4eoWJwAEK1G-GJaWG96jn5hXClgvpG8LpXDU0gLa5DjAoB0+sC5IMAG4QYC8OAABlUam6vX6gxIw3s40mMzmi2W2nWmx2ewORxOhQsEDAMGAhmQWxC3TgAGErrYElA7hksrongVimV3pV4HD5HVMXZPFSaXSDAyQpSfHhbKI3JCGaCNVs4CVprMlEsSJpCddtrt9uqGR44AAuLpQLCiClwZ4ACUS4hicAAFshIeIYDIAO42GDiABEMtp9MZHgjKDsyFscAgMB9iWDwFQ4mA8GDhhI8HUuYMXQz+Fsmo5iSCXTClEpYGAYjcdV+Qf+zncXmpMflcY8AG5TuIACLsm5clFZXj8l5vIA

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
type Foo = 'foo' | 'bar'

const schema = z.enum( [ 'foo', 'bar' ] ).optional()
    .pipe( z.custom<Foo>().optional() )
type Data = z.infer<typeof schema>
//   ^? Foo | undefined

console.log( schema.parse( 'foo' ) ) // foo
console.log( schema.parse( 'bar' ) ) // bar
console.log( schema.parse( undefined ) ) // undefined
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
